pkg/infra/baseAgent.go

```diff
@@ -262,75 +265,57 @@ func (a *Base) UpdateHPChange(change int) {
 	}
 }
 
-func (a *Base) utility() float64 {
-	// TODO: Improve utility calculation
-	// Currently just use ratio between food taken and food seen
-	// Somehow also factor
-	// Ideas: piecewise/bounded function with log, concave quadratic
+func (a *Base) updateUtility(foodRequested, foodTaken food.FoodType) {
+	// 1.) Determines the resources it has available, gi ∈[0,1]
+	// 2.) Determines its need for resources, qi ∈[0,1]
+	// 3.) Makes a provision of resources, pi ∈[0,1]
+	// 4.) Makes a demand for resources, di ∈[0,1]
+	// 5.) Receives an allocation of resources, ri ∈[0,1]
+	// 6.) Makes an appropriation of resources, r′i ∈[0,1]
+
+	// The total resources accrued at the end of a round is hence given by:
+	// Ri = r′i + (gi −pi)
+	// Which provides the utility of an agent, ui, as:
+	// ui = αi(qi) + βi(Ri −qi), if Ri ≥qi
+	// 		αi(Ri) −γi(qi −Ri), otherwise
+
+	// Hyperparameters that are chosen to make utility commensurate for all agents
+	// Requires that alpha > gamma > beta for diminishing returns + greed
+
+	// The weighting assigned to the resources exactly needed to survive
+	alpha := 0.2
+	// The weighting assigned to bonus resources that aren't needed
+	beta := 0.1
+	// The weighting assigned to not getting the resources that are needed
+	gamma := 0.18
+
+	// Available resources normalised by total possible available resources
+	foodAvailable := float64(a.CurrPlatFood()) / float64(a.tower.maxPlatFood)
+
+	// If an agent is not about to die, it effectively needs 0 food
+	foodToSurvive := 0.0
+	// If the agent is on the verge of dying, food is needed
+	if a.daysAtCritical == a.HealthInfo().MaxDayCritical {
+		foodToSurvive = 1.0
+	}
+	
```

Comment:
Is this strictly the case? Surely when it's just critical and not just on its dying day it can count as needing food?

Reply:
But you can still survive. The only time you literally need food is when you're critical and about to die

Reply:
Unless I'm mistaken - can you be critical and die, even if it's not the dying day?

Reply:
I would say it is only "needed" the very last day. This is also the way to match the specs, i.e., having agents oscillating between critical state and WeakLevel while spending more time in the critical state.

As a comment, note also that 1.0 is not really the food to survive, but rather ressourceNeeded/maxRessourceNeeded, so that we have q_i in [0,1]. The actual amount of food to survive is actually 2 with the current health parameters

Reply:
Hm, you might be right on this one actually - any food here would be good utility

Reply:
@mdubied I'm happy to go with Gowoon on this one?

Reply:
> I would say it is only "needed" the very last day. This is also the way to match the specs, i.e., having agents oscillating between critical state and WeakLevel while spending more time in the critical state.
> 
> The variable name foodToSurvive is actually not perfect (but ok I think)
> 
> 1.0 is not really the food to survive, but rather ressourceNeeded/maxRessourceNeeded, so that we have q_i in [0,1]. The actual amount of food to survive is actually 2 with the current health parameters

My only issue with this (that makes me want to go with Gowoon) is that taking 2 food while in critical would otherwise be seen as wasted utility. I think this is probably unfair, even though they can technically survive a bit longer. Agents shouldn't be penalised for returning to weak when in critical

Reply:
@MattScottEEE you can not die if it's not "the dying day" (what a sentence hahah)

I am not completely sure about it. I completely agree that the food at critical is good utility. But it we push the reasoning, this same amount of food is also good when we are at WeakLevel. With the same amount of food you can either go from ciritical to weak, or stay at weak if you already at weak. Therefore, I would say that the food is really "needed" only to avoid dying, i.e., when `DaysAtCritical=MaxDayCritical`

Reply:
If you're on day 2 at a critical state, you skip eating because day 3 is the last possible day and the platform gets to you on the next day (day 3) and it's empty (probably not an unusual case) - you're dead, you should have eaten on day 2 and you needed that food when it got to you.
This isn't exact because there are no guarantees but I think the critical cut-off is a good place to put the threshold and a useful distinction against idk `weak` for example.

Reply:
Hm, overall I think that we should be rewarding agents for taking food at the critical stage

How about a compromise:

Taking food when at the very final day rewards a q=1.0
Taking food when just at critical (before final day) rewards 0.5<q<1.0

@woonmoon @mdubied

Reply:
> If you're on day 2 at a critical state, you skip eating because day 3 is the last possible day and the platform gets to you on the next day (day 3) and it's empty (probably not an unusual case) - you're dead, you should have eaten on day 2 and you needed that food when it got to you.
> This isn't exact because there are no guarantees but I think the critical cut-off is a good place to put the threshold and a useful distinction against idk `weak` for example.

Whilst you may die from this, that doesn't really matter, as you'd technically be getting more utility for your food if you waited another day. It's just unlucky. At the same time, you definitely should get some utility for even waiting until critical. We could parameterise q as a mapping from days critical to 0->1 (daysCrit / maxDaysCrit?)

Reply:
It is really a good question to answer @woonmoon !
The argument you givis not completely satisfying in my opinion. Using the same reasoning, I could say that you need to eat already even earlier, for example when HP=20 to avoid the risky scenario where you get nothing on day 3 of being at critical. But we agree that eating when HP = 20 is not "needed"
 I quite like your compromise @MattScottEEE though

Reply:
@mdubied Yes but isn't this why we categorized HP into "how desperately do you need to eat"? Yes you could be unlucky and the platform could arrive empty at you for 20 days in a row and you're just dead. It's just not good utility for an agent to eat when they're healthy but the same can't be said if you know you're ~3 days out from dying. I agree with @MattScottEEE that we should not be penalizing agents for eating when they're critical.